akka-cluster-sharding/src/main/scala/akka/cluster/sharding/dynamic/DynamicShardAllocationStrategy.scala

```diff
@@ -0,0 +1,221 @@
+
+    override def preStart(): Unit = {
+      log.debug("Starting ddata state actor for [{}]", typeName)
+      DataKeys.foreach { key =>
+        replicator ! Subscribe(key, self)
+      }
+    }
+
+    var currentLocations: Map[ShardId, String] = Map.empty
+
+    override def receive: Receive = active
+
+    @silent("deprecated") // mapValues dance
+    private def active: Receive = {
+      case c @ Changed(key: LWWMapKey[ShardId, String] @unchecked) =>
+        val newLocations = c.get(key).entries
+        currentLocations ++= newLocations
+        log.debug("Received updated shard locations [{}] all locations are now [{}]", newLocations, currentLocations)
+      case GetShardLocation(shard) =>
+        log.debug("GetShardLocation [{}]", shard)
+        val shardLocation = currentLocations.get(shard).map(asStr => AddressFromURIString(asStr))
+        sender() ! GetShardLocationResponse(shardLocation)
+      case GetShardLocations =>
+        log.debug("GetShardLocations")
+        sender() ! GetShardLocationsResponse(currentLocations.mapValues(asStr => AddressFromURIString(asStr)).toMap)
+    }
+  }
+}
+
+class DynamicShardAllocationStrategy(system: ActorSystem, typeName: String)(
+    // local only ask
+    implicit val timeout: Timeout = Timeout(5.seconds))
+    extends ShardCoordinator.StartableAllocationStrategy {
+
+  import DynamicShardAllocationStrategy._
+  import akka.pattern.ask
+  import system.dispatcher
+
+  private val log = Logging(system, classOf[DynamicShardAllocationStrategy])
+
+  private var shardState: ActorRef = _
+
+  private[akka] def createShardStateActor(): ActorRef = {
+    system
+      .asInstanceOf[ExtendedActorSystem]
+      .systemActorOf(DDataStateActor.props(typeName), s"dynamic-allocation-state-$typeName")
+  }
+
+  private val cluster = Cluster(system)
+
+  override def start(): Unit = {
+    shardState = createShardStateActor()
+  }
+
+  override def allocateShard(
+      requester: ShardRegion,
+      shardId: ShardId,
+      currentShardAllocations: Map[ShardRegion, immutable.IndexedSeq[ShardId]]): Future[ShardRegion] = {
+
+    log.debug("allocateShard [{}] [{}] [{}]", shardId, requester, currentShardAllocations)
+
+    // current shard allocations include all current shard regions
+    (shardState ? GetShardLocation(shardId))
+      .mapTo[GetShardLocationResponse]
+      .map {
+        case GetShardLocationResponse(None) =>
+          log.debug("No specific location for shard [{}]. Allocating to requester [{}]", shardId, requester)
+          requester
+        case GetShardLocationResponse(Some(address)) =>
+          // if it is the local address, convert it so it is found in the shards
+          if (address == cluster.selfAddress) {
+            currentShardAllocations.keys.find(_.path.address.hasLocalScope) match {
+              case None =>
+                log.warning("unable to find local shard in currentShardAllocation. Using requester")
```

Comment:
In what kind of situations might this happen?
* the local shardregion has not registered itself to the coordinator yet, but ddata already knows there should be shards allocated to us.
* we are leaving but the ddata still thinks there should be shards allocated to us

Not sure how we could do better though..

Reply:
And any shards that end up in the wrong place will be fixed on the next rebalance